### PR TITLE
[WIP]: POC Ayah highlights

### PR DESCRIPTION
**Pull request by muhajirdev**

### CODE REVIEW

File: src/components/Verse/VerseText.tsx [hunk 1]
```diff
@@ -37,9 +47,23 @@ const VerseText = ({ words }: VerseTextProps) => {
...
       >
-        {words?.map((word) => (
-          <QuranWord key={word.location} word={word} font={quranReaderStyles.quranFont} />
-        ))}
+        {words?.map((word, index) => {
+          const highlight =
+            segments &&
+            // example: bismillahirrahmanirrahim, is detected as 4 words with chapterTypeName 'word'
+            // + 1 word with chapterTypeName 'end'. So 5 word in total. Need to check before doing shouldHighlight
+            word.charTypeName === 'word' &&
```
Comment (osamasayed) [inline src/components/Verse/VerseText.tsx:55]: You can use `CharType` enum for this check `CharType.Word`.

Comment (muhajirdev) [reply src/components/Verse/VerseText.tsx]: thanks

File: src/components/QuranReader/selectVerseHighlight.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  return wordPosition + 1; // +1 because the word position is not 0 based. while in timestampSegments it is 0 based
+};
+
+// to be used with `useSelector` to prevent unnecassary re rendering
+const selectHighlightedWordPosition = (state, timestampSegments: [number[]]) => {
+  const currentTime = selectCurrentTime(state);
+  return findHighlightedWordPosition(currentTime, timestampSegments);
+};
+
+const selectVerseHighlight = (state, verse: Verse) => ({
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlight.ts:41]: selectVerseHighlightStatus

File: src/components/QuranReader/TranslationView/index.tsx [hunk 3]
```diff
@@ -5,46 +5,71 @@ import Link from 'next/link';
...
+  verse,
+  quranReaderStyles,
+}: {
+  verse: Verse;
+  quranReaderStyles: QuranReaderStyles;
+}) => {
+  const { verseHighlighted: highlighted, wordPosition: position } = useSelector(
+    (state) => selectIsVerseHighlighted(state, verse),
+    shallowEqual,
+  );
```
Comment (tifa2UP) [inline src/components/QuranReader/TranslationView/index.tsx:33]: Rename, and move component to a separate file.

File: src/components/QuranReader/selectVerseHighlight.ts (hunk 2 above)
Comment (tifa2UP) [reply src/components/QuranReader/selectVerseHighlight.ts:41]: Add documentation to functions

File: src/components/QuranReader/selectVerseHighlight.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import { selectAudioFile, selectCurrentTime } from 'src/redux/slices/AudioPlayer/state';
+import Verse from 'types/Verse';
+import inRange from 'lodash/inRange';
+
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlight.ts:4]: // Returns whether the verse should be highlighted given the current audio timestamp

File: src/components/QuranReader/selectVerseHighlight.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+const selectIsVerseHighlighted = (state, verse: Verse) => {
+  const audioFile = selectAudioFile(state);
+  if (verse.chapterId !== audioFile.chapterId) return false;
+
+  const currentTime = selectCurrentTime(state);
+  const currentTimeInSeconds = currentTime * 1000;
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlight.ts:10]: currentTimeInMS

File: src/components/QuranReader/selectVerseHighlight.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    verse.timestamps.timestampTo,
+  );
+};
+
+const shouldHighlightWord = (currentTime: number, segment: number[]) => {
+  const startTime = segment[1];
+  const endTime = segment[2];
+  const currentTimeInMiliseconds = currentTime * 1000;
+  return inRange(currentTimeInMiliseconds, startTime, endTime);
+};
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlight.ts:23]: review the efficiency of this

File: src/components/Verse/VerseText.tsx [hunk 7]
```diff
@@ -43,7 +44,12 @@ const VerseText = ({ words, isReadingMode = false }: VerseTextProps) => {
           })}
         >
           {words?.map((word) => (
-            <QuranWord key={word.location} word={word} font={quranReaderStyles.quranFont} />
+            <QuranWord
+              key={word.location}
+              word={word}
+              font={quranReaderStyles.quranFont}
+              highlight={highlightedWordPosition === word.position}
```
Comment (tifa2UP) [inline src/components/Verse/VerseText.tsx:51]: renamed highlight to highlighted

File: src/components/QuranReader/TranslationView/index.tsx (hunk 3 above)
Comment (tifa2UP) [reply src/components/QuranReader/TranslationView/index.tsx:33]: isHighlighted for variables, highlighted for props is ok

Comment (tifa2UP) [reply src/components/QuranReader/TranslationView/index.tsx:33]: verseHighlighted -> isVerseHighlighted

File: src/components/QuranReader/TranslationView/index.tsx [hunk 8]
```diff
@@ -5,46 +5,71 @@ import Link from 'next/link';
...
 
 type TranslationViewProps = {
   verses: Verse[];
   quranReaderStyles: QuranReaderStyles;
 };
 
+// useSelector inside RenderVerse, for individual verse
+// instead of using useSelector in TranslationView
+// so only one verse get re rendered when the highlight changes
+const RenderVerse = ({
```
Comment (tifa2UP) [inline src/components/QuranReader/TranslationView/index.tsx:23]: rename to TranslationViewCell

File: src/components/QuranReader/selectVerseHighlight.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import { selectAudioFile, selectCurrentTime } from 'src/redux/slices/AudioPlayer/state';
+import Verse from 'types/Verse';
+import inRange from 'lodash/inRange';
+
+const selectIsVerseHighlighted = (state, verse: Verse) => {
+  const audioFile = selectAudioFile(state);
+  if (verse.chapterId !== audioFile.chapterId) return false;
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlight.ts:7]: if (!audioFile || verse.chapterId !== audioFile.chapterId)

File: src/components/QuranReader/selectVerseHighlight.ts (hunk 6 above)
Comment (muhajirdev) [reply src/components/QuranReader/selectVerseHighlight.ts]: added a todo for this

File: src/components/QuranReader/selectVerseHighlightStatus.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+// see which word should be highlighted, and then return the word position
+const selectHighlightedWordPosition = (state, timestampSegments: [number[]]) => {
+  const currentTime = selectCurrentTime(state);
+  return findHighlightedWordPosition(currentTime, timestampSegments);
+};
+
+// given the current audio timestamp, and the Verse data
+// we want to see if "this verse should be highlighted or not"
+// and "which word in this verse should be highlighted"
+// TODO: @muhajirdev, review efficiency of this function. Check if https://github.com/reduxjs/reselect can boost performance
```
Comment (tifa2UP) [inline src/components/QuranReader/selectVerseHighlightStatus.ts:54]: nit: remove comment

**Comment by muhajirdev** [comment]
optimizing performance (avoiding re-rendering) is still a little bit messy. 
Probably it's more if VerseText / each word can listen directly to redux. What `verseKey` and what `wordPosition` is currently highlighted. 

That way we don't need to check the changes from parent components, so there's no need to re-render its siblings.
But this might require changing data flow. Probably we should do that in another PR?

**Pull request closed without merging**